File: verlet.py
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as animation


def billiard(numberOfParticles, numberOfIterations, radius, dt, KK):
    pos = np.zeros((2, numberOfParticles, numberOfIterations+1))
    pos[0,:,0] = np.random.uniform(-radius, radius, numberOfParticles)
    pos[1,:,0] = np.random.uniform(-radius, radius, numberOfParticles)

    angles = np.random.uniform(-np.pi, np.pi)
    vx = np.cos(angles)
    vy = np.sin(angles)

    for i in range(numberOfIterations):
        distance = np.sqrt(pos[0,:,i]**2 + pos[1,:,i]**2)
        acceleration = np.zeros((2, numberOfParticles))

        #test if we are outside the circle and calculate force from wall
        acceleration[0, distance > radius] = -KK * (distance[distance > radius] - radius) * pos[0, distance > radius, i] / distance[distance > radius]
        acceleration[1, distance > radius] = -KK * (distance[distance > radius] - radius) * pos[1, distance > radius, i] / distance[distance > radius]

        #update positions
        pos[0, :, i+1] =  pos[0, :, i] + vx * dt + (acceleration[0, :] * dt**2)/2
        pos[1, :, i+1] =  pos[1, :, i] + vy * dt + (acceleration[1, :] * dt**2)/2

        distance = np.sqrt(pos[0,:,i+1]**2 + pos[1,:,i+1]**2)
        acceleration2 = np.zeros((2, numberOfParticles))
        #test if we are outside the circle and calculate force from wall
        acceleration2[0, distance > radius] = -KK * (distance[distance > radius] - radius) * pos[0, distance > radius, i+1] / distance[distance > radius]
        acceleration2[1, distance > radius] = -KK * (distance[distance > radius] - radius) * pos[1, distance > radius, i+1] / distance[distance > radius]

        #update velocities
        vx += (acceleration[0,:] + acceleration2[0,:])/2 * dt
        vy += (acceleration[1,:] + acceleration2[1,:])/2 * dt

    return pos


pos = billiard(1, 100000, 10, 0.005, 5)

plt.plot(pos[0,0,0:-1:100], pos[1,0,0:-1:100])
plt.xlim(-15,15)
plt.ylim(-15,15)
plt.show()

File: verlet.jl
function initial_values(numberOfParticles, radius)
    pos = rand((-radius:radius), 2, numberOfParticles)
    angles = rand((-pi:pi), numberOfParticles)
    return pos, angles
end


function billiard(
    numberOfParticles,
    numberOfIterations,
    radius,
    dt,
    KK;
    initial_value=nothing
)

    if (isnothing(initial_value))
        initial_value = initial_values(numberOfParticles, radius)
    end
    pos = zeros(2, numberOfParticles, numberOfIterations+1)
    velocity = Array{Float64, 3}(undef, 2, numberOfParticles, numberOfIterations+1)

    pos[:, :, 1], angles = initial_value

    vx = cos.(angles)
    vy = sin.(angles)
    velocity[:, :, 1] = transpose([vx vy])

    for i in 1:numberOfIterations, j in 1:numberOfParticles
        distance = sqrt(pos[1,j,i]^2 + pos[2,j,i]^2)
        if distance > radius
            acceleration = -KK * ((distance - radius)/distance) * pos[:, j, i]
        else
            acceleration = [0, 0]
        end

        pos[:, j, i+1] = pos[:, j, i] + velocity[:, j, i]*dt + acceleration*dt^2 / 2

        distance = sqrt(pos[1, j, i+1]^2 + pos[2, j, i+1]^2)
        if distance > radius
            acceleration2 = -KK * ((distance - radius)/distance) * pos[:, j, i+1]
        else
            acceleration2 = [0, 0]
        end

        velocity[:, j, i+1] = velocity[:, j, i] + (acceleration + acceleration2)/2 * dt
    end

    return pos, velocity
end

function potential(x, y, radius, KK)
    distance = sqrt(x^2 + y^2)
    if distance <= radius
        return 0
    end
    return (distance - radius)^2 * KK / 2
end

function energy(x, y, vel_x, vel_y, radius, KK)
    pot = potential(x, y, radius, KK)
    total_vel_square = vel_x^2 + vel_y^2

    return 0.5*total_vel_square + pot
end
